src/controller/java/src/matter/controller/MatterControllerImpl.kt

```diff
@@ -0,0 +1,661 @@
+ *   you may not use this file except in compliance with the License.
+ *   You may obtain a copy of the License at
+ *
+ *       http://www.apache.org/licenses/LICENSE-2.0
+ *
+ *   Unless required by applicable law or agreed to in writing, software
+ *   distributed under the License is distributed on an "AS IS" BASIS,
+ *   WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ *   See the License for the specific language governing permissions and
+ *   limitations under the License.
+ *
+ */
+package matter.controller
+
+import chip.devicecontroller.ChipDeviceController
+import chip.devicecontroller.ChipDeviceControllerException
+import chip.devicecontroller.GetConnectedDeviceCallbackJni.GetConnectedDeviceCallback
+import chip.devicecontroller.InvokeCallback
+import chip.devicecontroller.ReportCallback
+import chip.devicecontroller.ResubscriptionAttemptCallback
+import chip.devicecontroller.SubscriptionEstablishedCallback
+import chip.devicecontroller.WriteAttributesCallback
+import chip.devicecontroller.model.AttributeWriteRequest
+import chip.devicecontroller.model.ChipAttributePath
+import chip.devicecontroller.model.ChipEventPath
+import chip.devicecontroller.model.ChipPathId
+import chip.devicecontroller.model.InvokeElement
+import java.util.logging.Level
+import java.util.logging.Logger
+import kotlin.coroutines.resume
+import kotlin.coroutines.resumeWithException
+import kotlinx.coroutines.channels.Channel.Factory.UNLIMITED
+import kotlinx.coroutines.channels.awaitClose
+import kotlinx.coroutines.channels.onFailure
+import kotlinx.coroutines.channels.trySendBlocking
+import kotlinx.coroutines.flow.Flow
+import kotlinx.coroutines.flow.buffer
+import kotlinx.coroutines.flow.callbackFlow
+import kotlinx.coroutines.suspendCancellableCoroutine
+
+/** Controller to interact with the CHIP device. */
+class MatterControllerImpl(params: ControllerParams) :
+  MatterController, chip.devicecontroller.ChipDeviceController.CompletionListener {
+  private val deviceController: ChipDeviceController
+  private var completionListener: MatterController.CompletionListener? = null
+  private var nodeId: Long? = null
+
+  override fun setCompletionListener(listener: MatterController.CompletionListener) {
+    completionListener = listener
+  }
+
+  override fun pairDevice(
+    nodeId: Long,
+    address: String,
+    port: Int,
+    discriminator: Int,
+    pinCode: Long,
+  ) {
+    this.nodeId = nodeId
+    deviceController.setCompletionListener(this)
+    deviceController.pairDeviceWithAddress(nodeId, address, port, discriminator, pinCode, null)
+  }
+
+  override fun unpairDevice(nodeId: Long) {
+    deviceController.unpairDevice(nodeId)
+    this.nodeId = null
+  }
+
+  override fun establishPaseConnection(
+    nodeId: Long,
+    address: String,
+    port: Int,
+    setupPincode: Long
+  ) {
+    deviceController.establishPaseConnection(nodeId, address, port, setupPincode)
```

Comment:
Would this need a completion listener too? However I believe if we use an adapter we should not care.

Reply:
No, we don't need a separate one